### PR TITLE
Use non-crypto version of random to generate ids.

### PR DESCRIPTION
**Pull request by LK4D4**

### CODE REVIEW

File: pkg/random/random.go [hunk 1]
```diff
@@ -32,3 +39,10 @@ func NewSource() rand.Source {
 		src: rand.NewSource(time.Now().UnixNano()),
 	}
 }
+
+type nonSeedSource struct {
+	*lockedSource
+}
+
+func (r *nonSeedSource) Seed(seed int64) {
```
Comment (tiborvass) [inline pkg/random/random.go]: stdlib implements this: https://golang.org/src/math/rand/rand.go#L242

Comment (tiborvass) [inline pkg/random/random.go]: nvm, my bad

Comment (LK4D4) [inline pkg/random/random.go]: And that's why we use our own I think :)

File: pkg/stringid/stringid.go [hunk 2]
```diff
@@ -47,3 +51,29 @@ func GenerateRandomID() string {
...
 }
+
+// GenerateNonCryptoID generates unique id without using cryptographically
+// secure sources of random.
+// It helps you to save entropy.
+func GenerateNonCryptoID() string {
+	r := mrand.New(random.Source)
+	id := make([]byte, 32)
+	for {
+		if _, err := io.ReadFull(rand.Reader, id); err != nil {
```
Comment (tiborvass) [inline pkg/stringid/stringid.go]: this is still using crypto

File: pkg/stringid/stringid.go [hunk 3]
```diff
@@ -47,3 +51,29 @@ func GenerateRandomID() string {
...
+			panic(err) // This shouldn't happen
+		}
+		binary.PutUvarint(id, uint64(r.Int63()))
+		binary.PutUvarint(id[8:], uint64(r.Int63()))
+		binary.PutUvarint(id[16:], uint64(r.Int63()))
+		// because PutUvarint writing 9 bytes
+		binary.PutUvarint(id[23:], uint64(r.Int63()))
+		value := hex.EncodeToString(id)
```
Comment (tiborvass) [inline pkg/stringid/stringid.go]: starting from here, it should be factored out in a function reused by both GenerateNonCryptoID and GenerateRandomID

**Comment by tiborvass** [comment]
LGTM

**Comment by cpuguy83** [comment]
LGTM

@tiborvass This says design-review....

**Comment by ewindisch** [comment]
LGTM but we need a stronger seed for math.rand (working a PR)

**Comment by tiborvass** [comment]
@ewindisch why?

**Comment by ewindisch** [comment]
@tiborvass time-based seeds are predictable and easily brute-forced. I don't think our use of it is particularly unsafe here, but it costs little to do the initial seed from crypto/rand

See #15123

**Comment by stevvooe** [comment]
This is going to introduce a future security issue. The fact that IDs don't collide is a security feature, especially for image IDs.

We need to fix the entropy issue.

**Comment by tiborvass** [comment]
@stevvooe image IDs use crypto only.

**Comment by stevvooe** [comment]
Should I file the CVE now or should I wait for this to cause a problem?

**Comment by ewindisch** [comment]
I still contest there is no "entropy issue", but perhaps I digress.

@stevvooe we cannot guarantee that image IDs will not collide and cannot rely on that as a security feature. We have documented knowledge of why that's bad. Content-addressable IDs are the only IDs that may be relied on for distribution.

The use of math.rand in this patch *seems* to be safe and this PRNG is seeded by /dev/urandom.

**Comment by stevvooe** [comment]
@ewindisch @tiborvass This becomes a security issue when someone submits a PR with `stringid.GenerateNonCryptoID` in the wrong spot. At least with cryptographically strong random identifiers, we have some guarantee that they won't collide.

**Comment by ewindisch** [comment]
@stevvooe Certainly. My perspective is that this patch doesn't seem to do that in and of itself. While possibly harmless by itself, I do agree that this patch is unnecessary.

I still contend that if collision attacks are a threat, then UUID is the wrong mechanism. I have no problem with generating non-cryptographicly secure UUIDs.  UUIDs prevent casual collisions for applications which must provide identifiers. It is for applications where an increment counter would be safe and sufficient. UUIDs should be local, not for a multi-user, open distributed system. If UUIDs are used for any of these reasons then their use should be reassessed.

**Comment by ramm** [comment]
We just hit the containers ids (well, hostnames based on ids) collision on different machines in a swarm cluster. Caused a bit of trouble.

**Comment by justincormack** [comment]
Can you expand on that? What happened and what were the consequences?

**Comment by stevvooe** [comment]
@justincormack Speculating here, but if the hostnames (containerIDs) are exported to provide routing to specific containers, collisions will result in a random containers not being accessible. This problem is compounded by the fact that the hostname is a truncated container ID, so we lose even more entropy. With the short ids used as hostnames, this is only 48-bits.

I suspect the truncation is having a larger effect on collision probability than reduced entropy of prng used to generate ids. However, the [prng used for id generation](https://golang.org/src/math/rand/rng.go) isn't necessarily secure, so this could be even more likely than we think.

@ramm What is the observed rate of collision? Was this only a one off occurrence or are you seeing this regularly? If regularly, how many containers are being generated per day/hour/week?

**Comment by tiborvass** [comment]
I would suggest opening a new issue instead of commenting under this merged PR. Thanks!

**Pull request merged**